src/crypto/siphash.cpp

```diff
@@ -45,30 +48,57 @@ CSipHasher& CSipHasher::Write(uint64_t data)
     return *this;
 }
 
+
+/// Load a uint64_t from 0 to 7 bytes.
+inline uint64_t ReadU64ByLenLE(const unsigned char* data, size_t len)
+{
+    assert(len < 8);
+    uint64_t out = 0;
+    for (size_t i = 0; i < len; ++i) {
```

Comment:
On LE systems I can imagine that this is even faster:

```c++
uint64_t val = 0;
memcpy((unsigned char*)&val, data, len);
return val;
```

On BE it needs `return (val >> 8 * (8 - len))` instead.

Feel like benchmarking that?

Reply:
I'll benchmark, but I'm not too optimistic about this being faster, because a the length isn't statically known so it must actually call memcpy and can't inline and optimize.

I also want to add fuzz harnesses against reference implementations for the hash functions, which should increase confidence in refactoring the delicate areas(tough not sure how much reviewers will like the extra code)

Reply:
Benchmarked right now, they are too close to actually know which is faster

diff:
```C++
diff --git a/src/crypto/siphash.cpp b/src/crypto/siphash.cpp
index cfc04c194e..616bab40f1 100644
--- a/src/crypto/siphash.cpp
+++ b/src/crypto/siphash.cpp
@@ -54,9 +54,7 @@ inline uint64_t ReadU64ByLenLE(const unsigned char* data, size_t len)
 {
     assert(len < 8);
     uint64_t out = 0;
-    for (size_t i = 0; i < len; ++i) {
-        out |= (uint64_t)data[i] << (i * 8);
-    }
+    memcpy(&out, data, len);
     return out;
 }
```

Before:
```
$ ./src/bench/bench_bitcoin -filter="SipHash"

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|          676,729.00 |            1,477.70 |    0.0% |    4,250,201.00 |    1,624,300.00 |  2.617 |     125,015.00 |    0.0% |      0.01 | `SipHash`

$ ./src/bench/bench_bitcoin -filter="SipHash"

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|          676,510.00 |            1,478.17 |    0.1% |    4,250,201.00 |    1,623,700.00 |  2.618 |     125,015.00 |    0.0% |      0.01 | `SipHash`

$ ./src/bench/bench_bitcoin -filter="SipHash"

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|          676,427.00 |            1,478.36 |    0.0% |    4,250,201.00 |    1,623,600.00 |  2.618 |     125,015.00 |    0.0% |      0.01 | `SipHash`


```

After:
```
$ ./src/bench/bench_bitcoin -filter="SipHash"

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|          676,646.00 |            1,477.88 |    0.0% |    4,250,232.00 |    1,624,200.00 |  2.617 |     125,023.00 |    0.0% |      0.01 | `SipHash`

$ ./src/bench/bench_bitcoin -filter="SipHash"

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|          676,239.00 |            1,478.77 |    0.0% |    4,250,232.00 |    1,623,300.00 |  2.618 |     125,023.00 |    0.0% |      0.01 | `SipHash`

$ ./src/bench/bench_bitcoin -filter="SipHash"

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|          676,756.00 |            1,477.64 |    0.1% |    4,250,232.00 |    1,623,700.00 |  2.618 |     125,023.00 |    0.0% |      0.01 | `SipHash`
```

Reply:
Cool, thanks for testing.